src/main/java/com/fasterxml/jackson/core/util/BufferRecyclerPool.java

```diff
@@ -1,17 +1,288 @@
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that does not use
+     * any pool but simply creates new instances when necessary.
+     */
+    public static class NonRecyclingPool implements BufferRecyclerPool
+    {
+        private static final long serialVersionUID = 1L;
+
+        private static final BufferRecyclerPool SHARED = new NonRecyclingPool();
+
+        // No instances beyond shared one should be constructed
+        private NonRecyclingPool() { }
+
+        /**
+         * Accessor for the shared singleton instance; due to implementation having no state
+         * this is preferred over creating instances.
+         *
+         * @return Shared pool instance
+         */
+        public static BufferRecyclerPool shared() {
+            return SHARED;
+        }
+
+        // // // Actual API implementation
+        
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            // Could link back to this pool as marker? For now just leave back-ref empty
+            return new BufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler recycler) {
+            ; // nothing to do, there is no underlying pool
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that uses
+     * {@link ConcurrentLinkedDeque} for recycling instances.
+     */
+    public static class ConcurrentDequePool implements BufferRecyclerPool
+    {
+        private static final long serialVersionUID = 1L;
+
+        private static final BufferRecyclerPool SHARED = new ConcurrentDequePool();
+
+        private final transient Deque<BufferRecycler> pool;
+
+        // // // Life-cycle (constructors, factory methods)
+
+        protected ConcurrentDequePool() {
+            pool = new ConcurrentLinkedDeque<>();
+        }
+
+        /**
+         * Accessor for getting the globally shared singleton instance.
+         * Note that if you choose to use this instance,
+         * pool may be shared by many other {@code JsonFactory} instances.
+         *
+         * @return Shared pool instance
+         */
+        public static BufferRecyclerPool shared() {
+            return SHARED;
+        }
+
+        /**
+         * Accessor for creating and returning a new, non-shared pool instance.
+         *
+         * @return Newly constructed, non-shared pool instance
+         */
+        public static BufferRecyclerPool nonShared() {
+            return new ConcurrentDequePool();
+        }
+
+        // // // JDK serialization support
+
+        /**
+         * To avoid serializing pool contents we made {@code pool} {@code transient};
+         * to compensate, need to re-create proper instance using constructor.
+         */
+        protected Object readResolve() {
+            return new ConcurrentDequePool();
```

Comment:
This is where we'd need to choose between returning `SHARED` vs creating new one, if we serialized some sort of marker.

Reply:
The `readResolve()` method is typically used to implement the Singleton pattern, where the same object needs to be returned after deserialization. I'd rather return the `SHARED` instance here, otherwise you're silently duplicating the instances available after deserialization, or in other words you will have `JsonFactory` created before and after deserialization that should have the same `SHARED` instance but in reality have different pool instances.

More in general I don't understand the need of a "non-shared" pool instance (at least for these "default" pool implementations provided by us). Can you please clarify in which situation an user may want to have such "non-shared" instance? If there isn't any evident need, I'd rather keep the pools constructors private and only expose the shared instance.

Reply:
On `readResolve()`: it (or `writeReplace()`) can also be used to prevent serializing part of state, such as actual cached contents. Since JDK serialization does not call any constructors on deserialization (nor, I think, initialization statements? Or does it?), it is necessary to either;

* Clear out cache contents Map before write (writeReplace), to read back empty Map (etc)
* Mark Map field `transient`; use `readResolve()` to create properly initialized empty instance (with empty Map)